File: apps/web/lib/api/backtest.ts
/**
 * 研究室回测 API client · P1-4d(ADR 0038 · authed-only)。
 * 照 lib/api/bot-preset.ts 的 authed 范式:authHeaders(token) + 自定义 Error 类 + readDetail。
 *
 * 🔴 红线:纯研究记录的【只读展示 + 发起】· 绝不下单 / 撮合 / 真实交易 · market 锁 crypto。
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? '/api-proxy'
const PREFIX = '/api/v1/backtest'

export type BacktestStatus = 'pending' | 'done' | 'error'
export type BacktestPeriod = '1m' | '5m' | '15m' | '30m' | '1h' | '1d' | '1w'

export interface BacktestCreateRequest {
  symbol: string
  start: string // YYYY-MM-DD
  end: string // YYYY-MM-DD
  market?: 'crypto' // 当前仅 crypto(后端 Literal)
  period?: BacktestPeriod
  sma_fast?: number
  sma_slow?: number
  initial_cash?: number
  leverage?: number
  commission_rate?: number
  slippage_bps?: number
}

export interface BacktestCreateResponse {
  id: number
  status: BacktestStatus
}

export interface BacktestRunListItem {
  id: number
  symbol: string
  market: string
  period: string
  start_date: string
  end_date: string
  status: BacktestStatus
  created_at: string
}

/**
 * vibe calc_metrics 的 16 个标量指标(metrics.csv 单行)。
 * ★ 量纲契约(ADR 0040):本 interface 的收益/回撤/胜率类全是【比率】(0.1=10% · 前端 fmtPct ×100);
 *   与 Trade.return_pct(百分比数值)量纲不同 —— 同一引擎两套量纲,选错格式化函数差 100 倍。
 */
export interface BacktestMetrics {
  final_value: number // 绝对额(报价币 USDT)· fmtNum
  total_return: number // 比率(0.1=10% · P1-3 实测 -0.185=-18.5%)· fmtPct
  annual_return: number // 比率 · fmtPct
  max_drawdown: number // 比率(≤0)· fmtPct
  sharpe: number // 无量纲比值 · fmtRatio
  calmar: number // 无量纲比值 · fmtRatio
  sortino: number // 无量纲比值 · fmtRatio
  win_rate: number // 比率(0.55=55%)· fmtPct
  profit_loss_ratio: number // 无量纲比值 · fmtRatio
  profit_factor: number // 无量纲比值 · fmtRatio
  max_consecutive_loss: number // 整数计数 · fmtInt
  avg_holding_days: number // 天(推断+弱印证 · 见量纲审计报告)
  trade_count: number // 整数计数 · 完整回合(开+平 · 17 回合=34 条买卖腿)· fmtInt
  benchmark_return: number // 比率 · fmtPct
  excess_return: number // 比率 · fmtPct
  information_ratio: number // 无量纲比值 · fmtRatio
}

/** equity.csv 一行。数值列全是【比率/绝对额】(与 metrics 同侧,无百分比数值)。 */
export interface EquityPoint {
  timestamp: string // 日期串("2025-01-18" · 1d 粒度)
  equity: number // 绝对额(策略权益)
  drawdown: number // 比率 ≤0(真机实测 -0.0201=回撤2.01% · 前端 ×100)
  benchmark_equity: number // 绝对额(买入持有基准)
  ret: number // 比率(推断 · 前端未渲染 · 上屏前须真机核原值)
  active_ret: number // 比率(推断 · 前端未渲染 · 上屏前须真机核原值)
}

/** trades.csv 一行(一条买卖腿 · 17 回合=34 条)。★return_pct 量纲与 metrics 不同! */
export interface Trade {
  timestamp: string // 日期串
  code: string // 'BTCUSDT'(无斜杠 vibe code)
  side: string // 'buy'=开仓腿 / 'sell'=平仓腿(当前 SMA 纯多头;若放开做空策略此语义需重审)
  price: number // 绝对价 · fmtNum
  qty: number // base 币数量(当前裸渲染 · 长尾显示待打磨,见审计报告)
  reason: string // 可读原因('entry' 等)
  pnl: number // 绝对额(报价币 · 真机实测 -27547.76)· fmtNum · 开仓腿无意义显「—」
  holding_days: number // 天(推断+弱印证)
  return_pct: number // ★【百分比数值】(-2.76=-2.76% · 真机实测)· fmtPctNum 绝不 ×100(P1-4e.fix)
}

export interface BacktestRunResponse {
  id: number
  symbol: string
  market: string
  period: string
  start_date: string
  end_date: string
  params_json: Record<string, unknown>
  status: BacktestStatus
  metrics_json: BacktestMetrics | null
  equity_json: EquityPoint[] | null
  trades_json: Trade[] | null
  run_card_json: Record<string, unknown> | null
  error: string | null
  created_at: string
  updated_at: string
}

export class BacktestApiError extends Error {
  constructor(
    public readonly status: number,
    public readonly detail: string,
    /** 会员刀2:429 结构化 detail 原样保留 */
    public readonly rawDetail: unknown = null,
  ) {
    super(`BacktestApi ${status}: ${detail}`)
    this.name = 'BacktestApiError'
  }
}

async function readDetail(resp: Response): Promise<{ text: string; raw: unknown }> {
  try {
    const body = (await resp.json()) as { detail?: unknown }
    return {
      text: typeof body.detail === 'string' ? body.detail : `HTTP ${resp.status}`,
      raw: body.detail ?? null,
    }
  } catch {
    return { text: `HTTP ${resp.status}`, raw: null }
  }
}

function authHeaders(token: string): HeadersInit {
  return { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' }
}

/** POST /api/v1/backtest · 发起一次回测 → { id, status:"pending" }。 */
export async function createBacktest(
  token: string,
  body: BacktestCreateRequest,
): Promise<BacktestCreateResponse> {
  const r = await fetch(`${API_BASE}${PREFIX}`, {
    method: 'POST',
    headers: authHeaders(token),
    body: JSON.stringify(body),
  })
  if (!r.ok) {
    const d = await readDetail(r)
    throw new BacktestApiError(r.status, d.text, d.raw)
  }
  return (await r.json()) as BacktestCreateResponse
}

/** GET /api/v1/backtest · 本人回测历史(created_at 倒序)。 */
export async function listBacktests(
  token: string,
  limit = 50,
  signal?: AbortSignal,
): Promise<BacktestRunListItem[]> {
  const r = await fetch(`${API_BASE}${PREFIX}?limit=${limit}`, {
    headers: authHeaders(token),
    signal,
  })
  if (!r.ok) {
    const d = await readDetail(r)
    throw new BacktestApiError(r.status, d.text, d.raw)
  }
  return (await r.json()) as BacktestRunListItem[]
}

/** GET /api/v1/backtest/{id} · 单条 full-data(含 metrics/equity/trades/run_card)· 越权 404。 */
export async function getBacktest(
  token: string,
  id: number,
  signal?: AbortSignal,
): Promise<BacktestRunResponse> {
  const r = await fetch(`${API_BASE}${PREFIX}/${id}`, {
    headers: authHeaders(token),
    signal,
  })
  if (!r.ok) {
    const d = await readDetail(r)
    throw new BacktestApiError(r.status, d.text, d.raw)
  }
  return (await r.json()) as BacktestRunResponse
}

File: apps/api/app/schemas/backtest.py
"""研究室回测 API Pydantic 契约 · P1-4c.5 full-data(ADR 0038 · B 档报告 UI 取数自渲染)。

═══════════════════════════════════════════════════════════════════════════
🔴 红线:回测是【纯研究记录】的只读展示 + 发起 —— 绝不下单 / 撮合 / 余额 / 真实交易。
   全部端点 authed-only,按 current_user.id 过滤(越权取他人 run 返回 404)。
═══════════════════════════════════════════════════════════════════════════

- BacktestCreateRequest  · POST 请求体(照 service.BacktestParams · 确认报告 C8)。
- BacktestCreateResponse · POST 响应(run id + status · 供前端拿 id 轮询)。
- BacktestRunListItem    · GET 列表项(精简 · 不带重数据 equity/trades)。
- BacktestRunResponse    · GET 单条(full-data · 含 equity/trades/run_card)。
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# 与 service._PERIOD_TO_INTERVAL 的 key 对齐(确认报告 C8)· 不支持的 period 在此层 422。
BacktestPeriod = Literal["1m", "5m", "15m", "30m", "1h", "1d", "1w"]


class BacktestCreateRequest(BaseModel):
    """POST /api/v1/backtest 请求体 · 字段照 BacktestParams(symbol/start/end 必填,余可选带默认)。"""

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(min_length=1, max_length=64)
    start: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")  # YYYY-MM-DD
    end: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")  # YYYY-MM-DD
    market: Literal["crypto"] = "crypto"  # 当前仅 crypto(模型注释 · 确认报告 C4)
    period: BacktestPeriod = "1d"
    sma_fast: int = Field(default=5, ge=1, le=400)
    sma_slow: int = Field(default=20, ge=1, le=400)
    initial_cash: float = Field(default=1_000_000.0, gt=0)
    leverage: float = Field(default=1.0, gt=0, le=125)


class BacktestCreateResponse(BaseModel):
    """POST 响应 · 返回 run id + status(pending)供前端轮询 GET /{id}。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    id: int
    status: str


class BacktestRunListItem(BaseModel):
    """GET /api/v1/backtest 列表项 · 精简(不带重数据 equity/trades · 避免列表 payload 过大)。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    id: int
    symbol: str
    market: str
    period: str
    start_date: str
    end_date: str
    status: str
    created_at: datetime


class BacktestRunResponse(BaseModel):
    """GET /api/v1/backtest/{id} 单条 · full-data(含 equity/trades/run_card · B 档报告渲染)。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    id: int
    symbol: str
    market: str
    period: str
    start_date: str
    end_date: str
    params_json: dict[str, object]
    status: str
    metrics_json: dict[str, object] | None = None
    equity_json: list[dict[str, object]] | None = None
    trades_json: list[dict[str, object]] | None = None
    run_card_json: dict[str, object] | None = None
    error: str | None = None
    created_at: datetime
    updated_at: datetime
